### PR TITLE
Make `tpp.matmul` and `tpp.brgemm` ternary

### PR DESCRIPTION
**Pull request by chelini**
Make `tpp.matmul` and `tpp.brgemm` ternary at the memref level, meaning three input and one output operand. Update XSMM lowering for `tpp.matmul` and `tpp.brgemm`, providing custom dispatch and invoking calls as LIBXSMM does. A follow up PR will clean-up `xsmm.ternary` and provide more verification for xsmm dialect, that at the moment has zero verification. Similarly, aliasing check for `tpp.matmul` and `tpp.brgemm` for the right-most operand and the output will be added separately.

### CODE REVIEW

File: include/TPP/Dialect/Xsmm/XsmmOps.td [hunk 1]
```diff
@@ -9,8 +9,8 @@
 #ifndef TPP_XSMM_OPS
 #define TPP_XSMM_OPS
 
-include "XsmmDialect.td"
-include "XsmmAttr.td"
+include "TPP/Dialect/Xsmm/XsmmDialect.td"
+include "TPP/Dialect/Xsmm/XsmmAttr.td"
```
Comment (rengolin) [inline include/TPP/Dialect/Xsmm/XsmmOps.td:13]: Why is this necessary?

File: include/TPP/Dialect/Xsmm/XsmmOps.td [hunk 2]
```diff
@@ -140,6 +141,34 @@ def Xsmm_UnaryOp : Xsmm_Op<"unary", []> {
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// MatmulOp
+//===----------------------------------------------------------------------===//
+
+def Xsmm_MatmulOp : Xsmm_Op<"matmul"> {
+  let summary = "matmul call operation.";
+  let arguments = (ins Xsmm_DataType:$dataType, Variadic<XsmmMemRef>:$inputs);
```
Comment (rengolin) [inline include/TPP/Dialect/Xsmm/XsmmOps.td:150]: I don't think we should be using `Variadic` here, since the number of arguments doesn't change.

This may have less human-written boilerplate, but it passes the wrong idea to the compiler and to whoemever reads this later. It wouldn't be a lot more code to have a clear number of arguments for each op type.

File: include/TPP/Dialect/Xsmm/XsmmOps.td [hunk 3]
```diff
@@ -226,4 +256,34 @@ def Xsmm_UnaryDispatchOp : Xsmm_Op<"unary.dispatch",[Pure]> {
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// MatmulDispatchOp
+//===----------------------------------------------------------------------===//
+
+// TODO: all the `flags` should be a list.
+class Xsmm_MatmulAndBrgemmDispatchOp<string mnemonic, list<Trait> traits = []> :
```
Comment (rengolin) [inline include/TPP/Dialect/Xsmm/XsmmOps.td:264]: I'd call it `GEMMLike`?

File: include/TPP/Dialect/Xsmm/XsmmOps.td [hunk 4]
```diff
@@ -226,4 +256,34 @@ def Xsmm_UnaryDispatchOp : Xsmm_Op<"unary.dispatch",[Pure]> {
...
+  Xsmm_Op<mnemonic, !listconcat(traits, [Pure])> {
+    let arguments = (ins DenseI64ArrayAttr:$inputs,
```
Comment (rengolin) [inline include/TPP/Dialect/Xsmm/XsmmOps.td:266]: Would be good to have an explanation of what the attributes refer to. Also for the other ops.

File: runtime/XsmmRunnerUtils.cpp [hunk 5]
```diff
@@ -119,8 +119,8 @@ extern "C" void _mlir_ciface_xsmm_matmul_invoke(const libxsmm_datatype dtype,
 }
 
 extern "C" int64_t _mlir_ciface_xsmm_matmul_dispatch(
-    const libxsmm_datatype dtype, const bool isVNNI, int64_t m, int64_t n,
-    int64_t k, int64_t lda, int64_t ldb, int64_t ldc) {
+    const libxsmm_datatype dtype, int64_t m, int64_t n, int64_t k, int64_t lda,
+    int64_t ldb, int64_t ldc, const libxsmm_gemm_flags flag) {
```
Comment (rengolin) [inline runtime/XsmmRunnerUtils.cpp:123]:
```suggestion
    int64_t ldb, int64_t ldc, const libxsmm_gemm_flags flags) {
```

File: runtime/XsmmRunnerUtils.cpp [hunk 6]
```diff
@@ -134,13 +134,11 @@ extern "C" int64_t _mlir_ciface_xsmm_matmul_dispatch(
   libxsmm_blasint k_int = k;
 
   libxsmm_gemm_shape l_shape;
-  libxsmm_bitfield l_flags;
-  if (isVNNI) {
+  libxsmm_bitfield l_flags = LIBXSMM_GEMM_FLAG_NONE;
+  if (flag == LIBXSMM_GEMM_FLAG_VNNI_B) {
```
Comment (rengolin) [inline runtime/XsmmRunnerUtils.cpp:138]:
```suggestion
  if (flag && LIBXSMM_GEMM_FLAG_VNNI_B) {
```

File: runtime/XsmmRunnerUtils.cpp [hunk 7]
```diff
@@ -393,12 +390,9 @@ _mlir_ciface_xsmm_brgemm_dispatch(const libxsmm_datatype dtype, bool isVNNI,
   libxsmm_blasint stride_b = ldb * k * typeSize;
 
   libxsmm_gemm_shape l_shape;
-  libxsmm_bitfield l_flags;
-  if (isVNNI) {
+  libxsmm_bitfield l_flags = LIBXSMM_GEMM_FLAG_NONE;
+  if (flag == LIBXSMM_GEMM_FLAG_VNNI_B) {
```
Comment (rengolin) [inline runtime/XsmmRunnerUtils.cpp:394]:
```suggestion
  if (flag && LIBXSMM_GEMM_FLAG_VNNI_B) {
```

**Review by rengolin — COMMENTED** [review]
I think your title to the PR is backwards... You're moving `matmul` and `brgemm` away from `ternary` no?

File: runtime/XsmmRunnerUtils.cpp [hunk 8]
```diff
@@ -593,18 +587,18 @@ extern "C" int iree_xsmm_brgemm_dispatch(void *context, void *params,
...
     int64_t ldc;
+    const libxsmm_gemm_flags flag;
   } xsmm_brgemm_dispatch_t;
   xsmm_brgemm_dispatch_t *p = (xsmm_brgemm_dispatch_t *) params;
-  p->address = _mlir_ciface_xsmm_brgemm_dispatch((libxsmm_datatype) p->dtype,
-                                                 p->vnni, p->m, p->n, p->k,
-                                                 p->lda, p->ldb, p->ldc);
+  p->address =
+      _mlir_ciface_xsmm_brgemm_dispatch((libxsmm_datatype)p->dtype, p->m, p->n,
+                                        p->k, p->lda, p->ldb, p->ldc, p->flag);
```
[the remark this answers is not in the document]
Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp:601]: Currently is a single 'flag'. Do you still prefer 'flags'?

[the remark this answers is not in the document]
Comment (rengolin) [reply runtime/XsmmRunnerUtils.cpp:601]: Yes, because this will eventually be for all flags.

[the remark this answers is not in the document]
Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp:601]: ok

File: include/TPP/Dialect/Xsmm/XsmmOps.td (hunk 1 above)
Comment (chelini) [reply include/TPP/Dialect/Xsmm/XsmmOps.td:13]: Keep things consistent with MLIR and how we do in other files.

File: include/TPP/Dialect/Xsmm/XsmmOps.td (hunk 3 above)
Comment (chelini) [reply include/TPP/Dialect/Xsmm/XsmmOps.td]: Went for MatmulLike as the op is called Matmul

File: include/TPP/Dialect/Xsmm/XsmmOps.td (hunk 4 above)
Comment (chelini) [reply include/TPP/Dialect/Xsmm/XsmmOps.td]: added.

**Comment by alheinecke** [comment]
why doing it piece-meal? Let's add percision for at least in, out, compute. Transpose, etc.

File: runtime/XsmmRunnerUtils.cpp (hunk 7 above)
Comment (rengolin) [reply runtime/XsmmRunnerUtils.cpp]: We still need to check against the individual flag (as there may be more), so `==` won't do.

Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp]: Currently, this is fine as this is actually a single flag. We need to extend `Xsmm_GemmFlags` to accept at list of enums instead of a single 1. Unclear how to do in .td

File: runtime/XsmmRunnerUtils.cpp (hunk 6 above)
Comment (adam-smnk) [reply runtime/XsmmRunnerUtils.cpp]: You probably want bitwise AND here:
` if (flag & LIBXSMM_GEMM_FLAG_VNNI_B) {`
If it still complains about implicit conversion, you might need to throw in `== 1` to check if it is set.

Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp]: I will try to OR prior to conversion to func. Will take a bit of time.

Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp]: yes, agree. There is not OR atm.

Comment (rengolin) [reply runtime/XsmmRunnerUtils.cpp]: > You probably want bitwise AND here

D'oh! That's what I meant. Sorry. Updated the comment.

File: include/TPP/Dialect/Xsmm/XsmmOps.td (hunk 2 above)
Comment (chelini) [reply include/TPP/Dialect/Xsmm/XsmmOps.td:150]: Yes, but this would require custom builders as the assembly format dictates the builder signature. I can do it but this will PR will likely grow even more.

File: runtime/XsmmRunnerUtils.cpp (hunk 6 above)
Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp]: currently is a single flag

Comment (rengolin) [reply runtime/XsmmRunnerUtils.cpp]: I don;'t think the AND and the OR need to be done at the same time. You can easily check for the AND here and do the OR to set the flags later.

**Review by KavithaTipturMadhu — APPROVED** [review]
